File: sable-shared/src/logging/safeLog.ts
// PII-stripping log wrapper. The gateway DB doc requires "no PII in any log
// field" — every service routes log writes through `safeLog` so it's
// physically impossible for a sensitive field name to leak into stdout.
//
// Two exports:
//   - redact(value)   — pure function; returns a deeply-cloned value with
//                       sensitive fields replaced by '[REDACTED]'.
//   - safeLog(logger) — wraps any compatible Logger so its meta argument is
//                       redacted on every call.

import { isSensitiveField, normaliseField } from './sensitiveFields.js';

const REDACTED = '[REDACTED]';
const CIRCULAR = '[CIRCULAR]';
const DEPTH_EXCEEDED = '[DEPTH_EXCEEDED]';

// Patterns scanned inside string values when `scanStrings: true`. Default
// off — these regexes have a perf cost and can false-positive in URLs or
// non-PII text. Enable explicitly for error/stack-trace logging where the
// content is opaque and may contain inlined PII (Postgres errors that quote
// the offending row, third-party SDK error messages, etc.).
const EMAIL_RE = /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g;
const JWT_RE = /eyJ[A-Za-z0-9_-]+\.eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+/g;

function scanString(value: string): string {
  return value.replace(JWT_RE, '[JWT]').replace(EMAIL_RE, '[EMAIL]');
}

export interface RedactOptions {
  /** Extra field names to redact in addition to SENSITIVE_FIELDS. */
  extraFields?: Iterable<string>;
  /** Recursion cap. Default 10. */
  maxDepth?: number;
  /**
   * Scan inside string values for email and JWT patterns. Off by default
   * (cost + false-positive risk). Turn on for error / stack-trace logging.
   */
  scanStrings?: boolean;
}

// ---------------------------------------------------------------------------
// Core redactor
// ---------------------------------------------------------------------------

export function redact(value: unknown, options: RedactOptions = {}): unknown {
  const extras = options.extraFields
    ? new Set(Array.from(options.extraFields, normaliseField))
    : undefined;
  const seen = new WeakSet<object>();
  return walk(value, extras, seen, 0, options.maxDepth ?? 10, options.scanStrings ?? false);
}

function walk(
  value: unknown,
  extras: ReadonlySet<string> | undefined,
  seen: WeakSet<object>,
  depth: number,
  maxDepth: number,
  scanStrings: boolean,
): unknown {
  if (depth >= maxDepth) return DEPTH_EXCEEDED;
  if (value === null || value === undefined) return value;

  const t = typeof value;
  if (t === 'string') return scanStrings ? scanString(value as string) : value;
  if (t === 'number' || t === 'boolean' || t === 'bigint') return value;
  if (t === 'function' || t === 'symbol') return undefined;

  // Common opaque object types — emit a summary, never the contents.
  if (value instanceof Date) return value.toISOString();
  if (value instanceof Error) {
    return {
      name: value.name,
      message: value.message,
      // Stack is allowed in logs but redacted-stripped of email-like params.
      stack: value.stack,
    };
  }
  if (typeof Buffer !== 'undefined' && Buffer.isBuffer(value)) {
    return `[Buffer ${value.length} bytes]`;
  }
  if (ArrayBuffer.isView(value)) {
    const view = value as ArrayBufferView;
    return `[${view.constructor.name} ${view.byteLength} bytes]`;
  }

  if (typeof value === 'object') {
    if (seen.has(value)) return CIRCULAR;
    seen.add(value);

    if (Array.isArray(value)) {
      return value.map((v) => walk(v, extras, seen, depth + 1, maxDepth, scanStrings));
    }

    if (value instanceof Map) {
      const out: Record<string, unknown> = {};
      for (const [k, v] of value) {
        const key = String(k);
        out[key] = isSensitiveField(key, extras)
          ? REDACTED
          : walk(v, extras, seen, depth + 1, maxDepth, scanStrings);
      }
      return out;
    }

    if (value instanceof Set) {
      return Array.from(value, (v) => walk(v, extras, seen, depth + 1, maxDepth, scanStrings));
    }

    const out: Record<string, unknown> = {};
    for (const [key, val] of Object.entries(value as Record<string, unknown>)) {
      out[key] = isSensitiveField(key, extras)
        ? REDACTED
        : walk(val, extras, seen, depth + 1, maxDepth, scanStrings);
    }
    return out;
  }

  return value;
}

// ---------------------------------------------------------------------------
// Logger wrapper
// ---------------------------------------------------------------------------

export interface Logger {
  debug(message: string, meta?: unknown): void;
  info(message: string, meta?: unknown): void;
  warn(message: string, meta?: unknown): void;
  error(message: string, meta?: unknown): void;
}

/**
 * Wrap a Logger so every log call redacts its meta argument first. The
 * message itself is passed through unchanged — keep PII out of message
 * strings yourself; the wrapper only protects structured fields.
 */
export function safeLog(logger: Logger, options: RedactOptions = {}): Logger {
  return {
    debug: (message, meta) => logger.debug(message, meta === undefined ? meta : redact(meta, options)),
    info:  (message, meta) => logger.info(message, meta === undefined ? meta : redact(meta, options)),
    warn:  (message, meta) => logger.warn(message, meta === undefined ? meta : redact(meta, options)),
    error: (message, meta) => logger.error(message, meta === undefined ? meta : redact(meta, options)),
  };
}

File: sable-shared-py/sable_shared/log_safety/safe_log.py
"""PII-stripping log helpers.

The gateway DB doc requires "no PII in any log field" — every service
routes structured log meta through ``redact()`` so a sensitive field
name can't leak into stdout.

Two exports:
  - ``redact(value)``  — pure function; returns a deeply-cloned value
                          with sensitive fields replaced by ``[REDACTED]``.
  - ``safe_extra(meta)`` — convenience wrapper for use with stdlib
                          ``logging`` calls: ``log.info(msg, extra=safe_extra({...}))``.

Mirrors TS ``sable-shared/src/logging/safeLog.ts``.
"""

from __future__ import annotations

import datetime as _dt
import re
from typing import Any, Iterable

from sable_shared.log_safety.sensitive_fields import is_sensitive_field, normalise_field

REDACTED = "[REDACTED]"
CIRCULAR = "[CIRCULAR]"
DEPTH_EXCEEDED = "[DEPTH_EXCEEDED]"

# Patterns scanned inside string values when ``scan_strings=True``.
# Default off — these regexes have a perf cost and can false-positive in
# URLs or non-PII text. Enable explicitly for error / stack-trace logging
# where the content is opaque and may contain inlined PII.
_EMAIL_RE = re.compile(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}")
_JWT_RE = re.compile(r"eyJ[A-Za-z0-9_-]+\.eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+")


def _scan_string(value: str) -> str:
    return _EMAIL_RE.sub("[EMAIL]", _JWT_RE.sub("[JWT]", value))


def redact(
    value: Any,
    *,
    extra_fields: Iterable[str] | None = None,
    max_depth: int = 10,
    scan_strings: bool = False,
) -> Any:
    """Return a deeply-cloned ``value`` with sensitive fields replaced.

    Args:
        extra_fields: Field names to redact in addition to the canonical
            ``SENSITIVE_FIELDS``. Case / underscore variations are
            normalised internally.
        max_depth: Recursion cap. Default 10 — anything deeper becomes
            ``[DEPTH_EXCEEDED]``.
        scan_strings: When True, scan inside string values for email and
            JWT patterns. Off by default.
    """
    extras = frozenset(normalise_field(f) for f in extra_fields) if extra_fields else None
    seen: set[int] = set()
    return _walk(value, extras, seen, 0, max_depth, scan_strings)


def _walk(
    value: Any,
    extras: frozenset[str] | None,
    seen: set[int],
    depth: int,
    max_depth: int,
    scan_strings: bool,
) -> Any:
    if depth >= max_depth:
        return DEPTH_EXCEEDED
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float, complex)):
        return value
    if isinstance(value, str):
        return _scan_string(value) if scan_strings else value
    if isinstance(value, (bytes, bytearray, memoryview)):
        return f"[bytes {len(value)}]"
    if isinstance(value, _dt.datetime):
        return value.isoformat()
    if isinstance(value, _dt.date):
        return value.isoformat()
    if isinstance(value, BaseException):
        return {
            "type": value.__class__.__name__,
            "message": str(value),
        }

    if isinstance(value, dict):
        ref = id(value)
        if ref in seen:
            return CIRCULAR
        seen.add(ref)
        out_dict: dict[str, Any] = {}
        for k, v in value.items():
            key = str(k)
            if is_sensitive_field(key, extras):
                out_dict[key] = REDACTED
            else:
                out_dict[key] = _walk(v, extras, seen, depth + 1, max_depth, scan_strings)
        return out_dict

    if isinstance(value, (list, tuple, set, frozenset)):
        ref = id(value)
        if ref in seen:
            return CIRCULAR
        seen.add(ref)
        return [_walk(v, extras, seen, depth + 1, max_depth, scan_strings) for v in value]

    # Fallback for arbitrary objects — never log __dict__ directly because
    # it may contain sensitive attrs we don't know to redact.
    return repr(value)


def safe_extra(meta: dict[str, Any], **redact_options: Any) -> dict[str, Any]:
    """Redact ``meta`` for direct use as the stdlib logger ``extra=`` arg.

    Returns a flat dict the logger can drop into ``record.__dict__``;
    nested redaction still applies.
    """
    redacted = redact(meta, **redact_options)
    return redacted if isinstance(redacted, dict) else {"value": redacted}
